packages/atlas-service/src/main.ts

```diff
@@ -124,6 +142,92 @@ export class AtlasService {
+    // way from those events or otherwise, so to get the renewed token we listen
+    // to the refresh-succeeded event and then kick of the "refresh"
+    // programmatically to be able to get the actual tokens back and sync them
+    // to the service state
+    let refreshing = false;
+    this.oidcPluginLogger.on('mongodb-oidc-plugin:refresh-succeeded', () => {
+      // In case our call to REFRESH_TOKEN_CALLBACK somehow started another
+      // refresh token instead of just returning token form the plugin state, we
+      // short circuit if plugin logged a refresh-succeeded event and this
+      // listener got triggered
+      if (refreshing) {
+        return;
+      }
+      refreshing = true;
+      void this.plugin.mongoClientOptions.authMechanismProperties
+        // WARN: in the oidc plugin refresh callback is actually the same method
+        // as sign in, so calling it here means that potentially this can start
+        // an actual sign in flow for the user instead of just trying to refresh
+        // the token
+        .REFRESH_TOKEN_CALLBACK(
+          { clientId: this.clientId, issuer: this.issuer },
+          {
+            // Required driver specific stuff
+            version: 0,
+          }
+        )
+        .then((token) => {
+          this.token = token;
+          this.oidcPluginSyncedFromLoggerState = 'authenticated';
+          this.oidcPluginLogger.emit('atlas-service-token-refreshed');
+        })
+        .catch(() => {
+          this.token = null;
+          this.oidcPluginSyncedFromLoggerState = 'error';
+          this.oidcPluginLogger.emit('atlas-service-token-refresh-failed');
+        })
+        .finally(() => {
+          refreshing = false;
+        });
+    });
+  }
+
+  private static async maybeWaitForToken({
+    signal,
+  }: { signal?: AbortSignal } = {}) {
+    if (signal?.aborted) {
+      return;
+    }
+    // In cases where we ended up in expired state, we know that oidc-plugin
+    // will try to refresh the token automatically, we can wait for this process
+    // to finish before proceeding with a request
+    if (this.oidcPluginSyncedFromLoggerState === 'expired') {
+      try {
+        await Promise.race([
+          // We are using our own events here and not oidc plugin ones because
+          // after plugin logged that token was refreshed, we still need to run
+          // REFRESH_TOKEN_CALLBACK to get the actual token value in the state
+          once(this.oidcPluginLogger, 'atlas-service-token-refreshed'),
+          once(this.oidcPluginLogger, 'atlas-service-token-refresh-failed'),
+          new Promise((resolve) => {
+            signal?.addEventListener('abort', resolve, { once: true });
+          }),
+        ]);
+      } catch {
+        // In case token refresh failed, we jsut allow the request to happen so
+        // that it can fail with an error that will be propagated to the user
+      }
```

Comment:
Redundant, those are events, not methods
```suggestion
      await Promise.race([
        // We are using our own events here and not oidc plugin ones because
        // after plugin logged that token was refreshed, we still need to run
        // REFRESH_TOKEN_CALLBACK to get the actual token value in the state
        once(this.oidcPluginLogger, 'atlas-service-token-refreshed'),
        once(this.oidcPluginLogger, 'atlas-service-token-refresh-failed'),
        new Promise((resolve) => {
          signal?.addEventListener('abort', resolve, { once: true });
        }),
      ]);
```